src/System.Text.CaseFolding/System/Text/CaseFolding/SimpleCaseFolding.cs

```diff
@@ -0,0 +1,613 @@
+            ref char src = ref MemoryMarshal.GetReference(source);
+
+            int length = source.Length;
+
+            ref char refMapBelow5FF = ref MapBelow5FF[0];
+
+            // For char below 0x5ff use fastest 1-level mapping.
+            while (length != 0 && src <= MaxChar)
+            {
+                dst = Unsafe.Add(ref refMapBelow5FF, src);
+                src = ref Unsafe.Add(ref src, 1);
+                dst = ref Unsafe.Add(ref dst, 1);
+                length--;
+            }
+
+            if (length == 0)
+            {
+                return;
+            }
+
+            ref ushort refMapLevel1 = ref s_MapLevel1;
+            ref char refMapData = ref s_refMapData;
+
+            // We catch a char above 0x5ff.
+            // Process it with more slow two-level mapping.
+            while (length != 0 && !IsSurrogate(src))
+            {
+                ushort v1 =  Unsafe.Add(ref refMapLevel1, src >> 8);
+                char ch1 = Unsafe.Add(ref refMapData, v1 + (src & 0xFF));
+                if (ch1 == 0)
+                {
+                    ch1 = src;
+                }
+
+                dst = ch1;
+                src = ref Unsafe.Add(ref src, 1);
+                dst = ref Unsafe.Add(ref dst, 1);
+                length--;
+            }
+
+            if (length == 0)
+            {
+                return;
+            }
+
+            ref ushort refMapSurrogateLevel1 = ref s_refMapSurrogateLevel1;
+            ref var refMapSurrogateData = ref s_refMapSurrogateData;
+
+            while (length != 0)
+            {
+                // We catch a high or low surrogate.
+                // Process it and fallback to fastest options.
+                char c1 = src;
+                var isHighSurrogateA = IsHighSurrogate(c1);
+
+                if (isHighSurrogateA)
+                {
+                    // Get low surrogates.
+                    length--;
+                    if (length == 0)
+                    {
+                        // No low surrogate - throw?
+                        // We would have to throw but we can do simple binary copy
+                        // because the simple case folding is only used for comparisons.
+                        dst = c1;
+                        return;
+                    }
+
+                    src = ref Unsafe.Add(ref src, 1);
+                    char c1Low = src;
+
+                    if (!IsLowSurrogate(c1Low))
+                    {
+                        // The index is Utf32 minus 0x10000 (UNICODE_PLANE01_START)
+                        var index1 = ((c1 - HIGH_SURROGATE_START) * 0x400) + (c1Low - LOW_SURROGATE_START);
+
+                        if (index1 <= 0xFFFF)
+                        {
+                            ushort v1 = Unsafe.Add(ref refMapSurrogateLevel1, index1 >> 8);
+                            char ch1 = Unsafe.Add(ref refMapSurrogateData, v1 + (index1 & 0xFF));
```

Comment:
This is a (high surrogate, low surrogate) tuple, not `char`.

Reply:
Fixed.